File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.js
function solution(record) {
    
    const id = new Map();
    const rec = [];
    
    for(let i=0; i<record.length; i++){
        const input = record[i].split(" ");
        if(input[0] != "Change"){
            rec.push([input[0], input[1]]);
        }
        
        if(input[0] != "Leave"){
            id.set(input[1], input[2]);
        }
    }
    
    var answer = [];
    
    for(let i=0; i<rec.length; i++){
        let output = rec[i];
        if(output[0] == "Enter"){
            answer.push(id.get(output[1]) + "님이 들어왔습니다.")
        }else{
            answer.push(id.get(output[1]) + "님이 나갔습니다.")
        }
    }
    
    
    return answer;
}

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;

class Solution {
    public String[] solution(String[] record) {
        
        HashMap<String, String> id = new HashMap<>();
        Deque<String[]> rec = new ArrayDeque<>();
        
        for(int i=0; i<record.length; i++){
            String[] input = record[i].split(" ");
            if(!input[0].equals("Change")){
                rec.add(new String[] {input[0], input[1]});
            }
            
            if(!input[0].equals("Leave")){
                id.put(input[1], input[2]);
            }
        }
        
        String[] answer = new String[rec.size()];
        
        for(int i=0; i<answer.length; i++){
            String output[] = rec.poll();
            if(output[0].equals("Enter")){
                answer[i] = id.get(output[1]) + "님이 들어왔습니다.";
            }else{
                answer[i] = id.get(output[1]) + "님이 나갔습니다.";
            }
        }
        
        return answer;
    }
}